src/test/java/uk/nhs/hee/tis/trainee/credentials/service/CredentialEventServiceTest.java

```diff
@@ -0,0 +1,169 @@
+ */
+
+package uk.nhs.hee.tis.trainee.credentials.service;
+
+import static org.hamcrest.Matchers.hasItem;
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.when;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import uk.nhs.hee.tis.trainee.credentials.dto.CredentialType;
+import uk.nhs.hee.tis.trainee.credentials.model.CredentialMetadata;
+import uk.nhs.hee.tis.trainee.credentials.repository.CredentialMetadataRepository;
+
+class CredentialEventServiceTest {
+
+  private static final UUID PLACEMENT_1_credentialId =
+      UUID.fromString("e2a26a50-2c51-4327-8692-1cb05b80af07");
+  private static final String PLACEMENT_1_traineeId = "47165";
+  private static final String PLACEMENT_1_tisId = "2bd3f009-9c7a-4352-bcc6-9dff8eac07d6";
+  private static final Instant PLACEMENT_1_issuedAt =
+      Instant.parse("2023-09-27T09:37:48.000+00:00");
+
+  private static final UUID PLACEMENT_2_credentialId =
+      UUID.fromString("e5c09a44-39ff-442f-b356-8424caa581d0");
+  private static final String PLACEMENT_2_traineeId = "47165";
+  private static final String PLACEMENT_2_tisId = "2bd3f009-9c7a-4352-bcc6-9dff8eac07d6";
+  private static final Instant PLACEMENT_2_issuedAt =
+      Instant.parse("2023-08-14T14:12:54.000+00:00");
+
+  private static final UUID PLACEMENT_3_credentialId =
+      UUID.fromString("6ada9134-a582-44a2-8ce0-106ae8b9bcd4");
+  private static final String PLACEMENT_3_traineeId = "47165";
+  private static final String PLACEMENT_3_tisId =
+      "72055636-1439-11ee-be56-0242ac120002";
+  private static final Instant PLACEMENT_3_issuedAt =
+      Instant.parse("2023-09-27T09:41:35.000+00:00");
+
+  private static final UUID PROGRAMME_1_credentialId =
+      UUID.fromString("6ada9134-a582-44a2-8ee0-125ae8b9bcd4");
+  private static final String PROGRAMME_1_traineeId = "47165";
+  private static final String PROGRAMME_1_tisId = "72055636-1439-11ee-be56-0242ac120002";
+  private static final Instant PROGRAMME_1_issuedAt =
+      Instant.parse("2023-09-27T09:41:35.000+00:00");
+
+  private static final UUID PROGRAMME_2_credentialId =
+      UUID.fromString("6ada9134-a582-44a2-8de0-105ae8b9bcd4");
+  private static final String PROGRAMME_2_traineeId = "47165";
+  private static final String PROGRAMME_2_tisId = "72055636-1439-11ee-be56-0242ac120002";
+  private static final Instant PROGRAMME_2_issuedAt =
+      Instant.parse("2023-08-27T09:41:35.000+00:00");
+
+  private CredentialMetadata placement1;
+  private CredentialMetadata placement2;
+  private CredentialMetadata placement3;
+  private CredentialMetadata programme1;
+  private CredentialMetadata programme2;
```

Comment:
I still find this very difficult to follow, check the previous comment for a code snippet that would clean this up a lot.